exp-models/tests/unit/utils/csv-writer-test.js

```diff
@@ -0,0 +1,160 @@
+            },
+            echo: {
+                foxtrot: {
+                    golf: four
+                }
+            },
+            hotel: {
+                india: five,
+                juliet: {
+                    kilo: six
+                }
+            },
+            lima: {
+                mike: {
+                    november: {
+                        oscar: {
+                            papa: {
+                                quebec: seven
+                            }
+                        }
+                    }
+                }
+            }
+        }),
+        {
+            alpha: one,
+            bravo: two,
+            'charlie.delta': three,
+            'echo.foxtrot.golf': four,
+            'hotel.india': five,
+            'hotel.juliet.kilo': six,
+            'lima.mike.november.oscar.papa.quebec': seven
+        }
+    );
+});
+
+test('squash with map', function(assert) {
+    assert.deepEqual(
+        [
+            {
+                alpha: one,
+                bravo: {
+                    charlie: one
+                }
+            },
+            {
+                alpha: two,
+                bravo: {
+                    charlie: two
+                }
+            },
+            {
+                alpha: three,
+                bravo: {
+                    charlie: three
+                }
+            }
+        ].map(squash),
+        [
+            {
+                alpha: one,
+                'bravo.charlie': one
+            },
+            {
+                alpha: two,
+                'bravo.charlie': two
+            },
+            {
+                alpha: three,
+                'bravo.charlie': three
+            }
+        ]
+    );
+});
+
+test('uniqueFields', function(assert) {
+    assert.deepEqual(
+        uniqueFields([
+            {
+                one
+            },
+            {
+                one,
+                two
+            },
+            {
+                one,
+                two,
+                three
+            }
+        ]),
+        [
+            one,
+            two,
+            three
+        ]
+    );
+});
+
+test('writeCSV', function(assert) {
+    const data = [
+        {
+            test: one
+        },
+        {
+            test: two
+        }
+    ];
+
+    assert.strictEqual(
+        writeCSV(data, ['test']),
+        `test${NL}"one"${NL}"two"`
+    );
+});
+
+test('All together', function(assert) {
+    const data = [
+        {
+            a: one,
+            mic: {
+                check: two
+            }
+        },
+        {
+            a: one,
+            mic: {
+                check: two
+            }
+        }
+    ];
+
+    const squashedData = data.map(squash);
+    const fields = uniqueFields(squashedData);
+    const result = writeCSV(squashedData, fields);
+
+    assert.strictEqual(
+        result,
+        `a,mic.check${NL}"one","two"${NL}"one","two"`
+    );
+});
```

Comment:
Good tests so far!

Because this is schemaless data, the presence, number, and type of fields in a given record isn't really guaranteed- which admittedly is an unusual challenge for a CSV file!

There are a few edge cases where tests might be helpful to catch situations that we could reasonably expect to happen in production. Would it be ok to ask for the following tests? (and any others you can think of along these lines)

1. Serialize 2-3 records that have some, but not all, of the same fields as each other. Make sure the gaps are handled correctly, and that the correct fields line up across rows.
2. Serialize 2-3 records with a pre-provided list of headers, including some fields that are not present in one or more of the given records.
3. Serialize 2-3 records that have different *kinds* of data in each record. For example, where a record was sometimes a leaf entry (null), but other times was an object or an array. This test is important because the `JSON.stringify` representation of the latter cases would break the CSV file, unless the whole thing was deliberately coerced to a string.  Serializing an entire variable-length array into a single cell of a spreadsheet isn't ideal, but it's about the best compromise we've got when dealing with the restrictions of a tabular format.